[textinput] implement forward history search

ferdymercury opened the pull request:
# This Pull request:

## Changes or fixes:
Adds forward history search support.

## Checklist:

- [x] tested changes locally
- [x] updated the docs (if necessary)

This PR fixes https://github.com/root-project/root/issues/10137

ferdymercury commented:
It's still not working, somehow it gets stuck when going forward.

jalopezg-r00t commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -116,11 +123,23 @@ namespace textinput {
-        }
+      if (fMode == kHistRevSearchMode)
+      {
+         for (size_t i = startAt, n = Hist->GetSize(); i < n; ++i) {
+           if (Hist->GetLine(i).find(fSearch) != std::string::npos) {
+             NewHistEntry = i;
+             break;
+           }
+         }
+      }
+      else
+      {
+         for (long i = startAt; i >= 0 ; --i) {
+            if (Hist->GetLine(i).find(fSearch) != std::string::npos) {
+               NewHistEntry = i;
+               break;
+            }
+         }
```
Same here.  I have the feeling that these two loops can be merged (being the iteration order the only thing that changes).  :slightly_smiling_face:

jalopezg-r00t commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -167,9 +186,10 @@ namespace textinput {
+    if (fMode == kHistRevSearchMode || fMode == kHistFwdSearchMode) {
       fSearch += C;
-      SetReverseHistSearchPrompt(R.fDisplay);
+      if (fMode == kHistFwdSearchMode) SetForwardHistSearchPrompt(R.fDisplay);
+      else SetReverseHistSearchPrompt(R.fDisplay);
```
With the above change this becomes
```suggestion
      SetHistSearchModePrompt(R.fDisplay);
```

jalopezg-r00t commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -243,11 +263,12 @@ namespace textinput {
         if (fSearch.empty()) return kPRError;
         fSearch.erase(fSearch.length() - 1);
-        SetReverseHistSearchPrompt(R.fDisplay);
+        if (fMode == kHistFwdSearchMode) SetForwardHistSearchPrompt(R.fDisplay);
+        else SetReverseHistSearchPrompt(R.fDisplay);
```
And again,
```suggestion
        SetHistSearchModePrompt(R.fDisplay);
```

jalopezg-r00t commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -262,10 +283,23 @@ namespace textinput {
         } else {
           ++fCurHistEntry;
         }
+        SetReverseHistSearchPrompt(R.fDisplay);
```
I think we don't need this, as the `[bkw'...']` prompt doesn't need updating if the user just pressed Ctrl+R.

ferdymercury replied:
But wouldn't it need updating from "fwd" to "bkw"  if I switch from CTRL+S to CTRL+R ?

ferdymercury replied:
Yes, I tried removing that part. If press two times CTRL+R and twice CTRL+S, I get:
```
[bkw'int'] int c
[bkw'int'] int b
[bkw'int'] int a
[bkw'int'] int b
[bkw'int'] int c
```

instead of `fwd` in the last two lines

jalopezg-r00t commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -262,10 +283,23 @@ namespace textinput {
+      } else if (M == kCmdForwardSearch) {
+         // Search again. Move to newer hist entry:
+         size_t prevHistEntry = fCurHistEntry;
+         if (fCurHistEntry == 0) {
```
Given that `fCurHistEntry` might also be `-1`, I think this should be
```suggestion
         if (fCurHistEntry <= (size_t)0) {
```

jalopezg-r00t commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -262,10 +283,23 @@ namespace textinput {
+            return kPRError;
+         }
+         --fCurHistEntry;
+         SetForwardHistSearchPrompt(R.fDisplay);
```
```suggestion
```

jalopezg-r00t commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -430,12 +464,20 @@ namespace textinput {
         PushUndo();
-        fMode = kHistSearchMode;
-        fSearch.clear();
+        if (fMode != kHistFwdSearchMode) fSearch.clear();
```
We can just use `fSearch.clear();`, can't we?

I think we don't reach here if `fMode == kHist{Rev,Fwd}SearchMode` (see lines 287-289 and lines 298-300).

ferdymercury replied:
good point

jalopezg-r00t commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -430,12 +464,20 @@ namespace textinput {
         if (UpdateHistSearch(R)) return kPRSuccess;
         return kPRError;
+      case kCmdForwardSearch:
+         PopUndo();
```
I think we should just `PushUndo();`, as in `kCmdReverseSearch`.

jalopezg-r00t commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -430,12 +464,20 @@ namespace textinput {
         return kPRError;
+      case kCmdForwardSearch:
+         PopUndo();
+         if (fMode != kHistRevSearchMode) fSearch.clear();
+         fMode = kHistFwdSearchMode;
+         SetForwardHistSearchPrompt(R.fDisplay);
+         fContext->GetKeyBinding()->EnableEscCmd(true);
+         if (UpdateHistSearch(R)) return kPRSuccess;
+         return kPRError;
```
After changing `SetForwardHistSearchPrompt()` to `SetHistSearchModePrompt()` the code for this case should be similar to `kCmdReverseSearch`.  Therefore, I'd suggest to merge both code paths as in:
```c++
case kCmdReverseSearch:
case kCmdForwardSearch:
  // ...
  fMode = (M == kCmdReverseSearch ? kHistRevSearchMode : kHistFwdSearchMode);
  // ...
  break;
```

jalopezg-r00t commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -549,6 +591,13 @@ namespace textinput {
                                       fContext->GetCursor()));
   }
 
+  void
+  Editor::PopUndo() {
+     if (!fUndoBuf.empty()) {
+        fUndoBuf.pop_back();
+     }
+  }
+
```
Do we really need this?

jalopezg-r00t commented on core/textinput/src/textinput/Editor.h:
```diff
@@ -140,26 +141,29 @@ namespace textinput {
     EProcessResult ProcessCommand(ECommandID M, EditorRange& R);
     size_t FindWordBoundary(int Direction);
     void PushUndo();
+    void PopUndo();
```
```suggestion
```

jalopezg-r00t commented on core/textinput/src/textinput/Editor.h:
```diff
@@ -140,26 +141,29 @@ namespace textinput {
     void AddToPasteBuf(int Dir, const std::string& T);
     void AddToPasteBuf(int Dir, char T);
     void ClearPasteBuf() { fCutDirection = 0; }
     void SetReverseHistSearchPrompt(Range& RDisplay);
+    void SetForwardHistSearchPrompt(Range& RDisplay);
```
See comments above.

ferdymercury commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -104,6 +104,13 @@ namespace textinput {
     RDisplay.ExtendPromptUpdate(Range::kUpdateEditorPrompt);
   }
 
+  void
+  Editor::SetForwardHistSearchPrompt(Range& RDisplay) {
+     std::string P("[fwd'");
+     SetEditorPrompt(Text(P + fSearch + "'] "));
+     RDisplay.ExtendPromptUpdate(Range::kUpdateEditorPrompt);
+  }
+
```
Good point, I'll try this now

jalopezg-r00t commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -90,40 +90,41 @@ namespace textinput {
-    std::string P("[bkw'");
-    SetEditorPrompt(Text(P + fSearch + "'] "));
+  Editor::SetHistSearchModePrompt(Range& RDisplay) {
+    const std::string direction(fMode == kHistFwdSearchMode ? "fwd" : fMode == kHistRevSearchMode ? "bkw" : "NA");
```
```suggestion
    assert(fMode == kHistFwdSearchMode || fMode == kHistRevSearchMode);
    const std::string direction(fMode == kHistFwdSearchMode ? "fwd" : "bkw");
```

jalopezg-r00t commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -90,40 +90,41 @@ namespace textinput {
         startAt = 0;
       }
-      for (size_t i = startAt, n = Hist->GetSize(); i < n; ++i) {
+      const bool forward = fMode == kHistFwdSearchMode;
+      for (long i = startAt ; forward ? i >= 0 : i < static_cast<long>(Hist->GetSize()); forward ? --i : ++i) {
```
The suggestion below is not tested, but it might be slightly more efficient.
```suggestion
      const ssize_t stopAt = (fMode == kHistFwdSearchMode ? -1 : static_cast<ssize_t>(Hist->GetSize());
      const ssize_t step = (fMode == kHistFwdSearchMode ? -1 : 1);
      for (ssize_t i = startAt; i != stopAt; i += step) {
```

jalopezg-r00t commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -243,29 +244,51 @@ namespace textinput {
         if (UpdateHistSearch(R)) return kPRSuccess;
         return kPRError;
       } else if (M == kCmdReverseSearch) {
+        if (previousMode != kHistRevSearchMode) { // it was in fwd mode, just update label
+          fMode = kHistRevSearchMode;
+          SetHistSearchModePrompt(R.fDisplay);
+          return kPRSuccess;
+        }
```
See above.
```suggestion
```

jalopezg-r00t commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -243,29 +244,51 @@ namespace textinput {
     }
-    if (fMode == kHistSearchMode) {
+    if (fMode == kHistRevSearchMode || fMode == kHistFwdSearchMode) {
+      const auto previousMode = fMode;
```
We can merge the `if` statements in lines 256-260 and lines 276-280 here.
```suggestion
      const auto previousMode = fMode;
      fMode = (M == kCmdReverseSearch ? kHistRevSearchMode : kHistFwdSearchMode);
      if (previousMode != fMode) { // changed from fwd to bkw or viceversa, just update label
        SetHistSearchModePrompt(R.fDisplay);
        return kPRSuccess;
      }
```

jalopezg-r00t commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -243,29 +244,51 @@ namespace textinput {
         return kPRError;
-      } else {
+      } else if (M == kCmdForwardSearch) {
+         if (previousMode != kHistFwdSearchMode) { // it was in bkw mode, just update label
+           fMode = kHistFwdSearchMode;
+           SetHistSearchModePrompt(R.fDisplay);
+           return kPRSuccess;
+         }
```
Same.  See above.
```suggestion
```

jalopezg-r00t reviewed (changes requested):
Almost there!  :slightly_smiling_face:

Please, test the implementation after applying the suggestions; after that, I think we can squash into a single commit and merge.

ferdymercury commented:
> Almost there! slightly_smiling_face
> 
> Please, test the implementation after applying the suggestions; after that, I think we can squash into a single commit and merge.

It's working well, but we have a crash if I perform the following sequence:

- press CTRL+S
- write 'int'
- press CTRL+S again

```
===========================================================
#10 __memchr_avx2 () at ../sysdeps/x86_64/multiarch/memchr-avx2.S:65
#11 0x00007f9a4575a505 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::find(char const*, unsigned long, unsigned long) const () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#12 0x00007f9a45e00162 in textinput::Editor::UpdateHistSearch (this=0x560b564df640, R=...) at /opt/root_src/core/textinput/src/textinput/Editor.cpp:124
#13 0x00007f9a45e00a77 in textinput::Editor::ProcessCommand (this=0x560b564df640, M=textinput::Editor::kCmdForwardSearch, R=...) at /opt/root_src/core/textinput/src/textinput/Editor.cpp:285
#14 0x00007f9a45dffb48 in textinput::Editor::Process (this=0x560b564df640, cmd=..., R=...) at /opt/root_src/core/textinput/src/textinput/Editor.cpp:67
#15 0x00007f9a45e1049d in textinput::TextInput::ProcessNewInput (this=0x7f9a462fd880 <(anonymous namespace)::TextInputHolder::getHolder()::sTIHolder>, in=..., R=...) at /opt/root_src/core/textinput/src/textinput/TextInput.cpp:174
#16 0x00007f9a45e10117 in textinput::TextInput::ReadInput (this=0x7f9a462fd880 <(anonymous namespace)::TextInputHolder::getHolder()::sTIHolder>) at /opt/root_src/core/textinput/src/textinput/TextInput.cpp:126
#17 0x00007f9a45dfe363 in Getlinem (mode=kOneChar, prompt=0x0) at /opt/root_src/core/textinput/src/Getline.cxx:256
===========================================================
```

ferdymercury commented:
Also, if I just press CTRL+R once, or CTRL+S once, before starting to type anything, it loads the last entry in the history. I think it shouldn't load anything.

ferdymercury commented:
> It's working well, but we have a crash if I perform the following (stupid) sequence:
> 
>     * press CTRL+S
> 
>     * write 'int'
> 
>     * press CTRL+S again

This is now solved.

ferdymercury commented:
> Also, if I just press CTRL+R once, or CTRL+S once, before starting to type anything, it loads the last entry in the history. I think it shouldn't load anything.

This is solved, too.

ferdymercury commented:
Remaining issues:

- if when searching in reverse, I press "Esc", the prompt updates to 'fwd'. I have to press Esc a second time to cancel
- if when searching, I press "Enter", instead of selecting the line, as it was the case before, it directly executes it.

ferdymercury commented:
> Remaining issues:
> 
>     * if when searching in reverse, I press "Esc", the prompt updates to 'fwd'. I have to press Esc a second time to cancel
> 
>     * if when searching, I press "Enter", instead of selecting the line, as it was the case before, it directly executes it.

both fixed now, too.

From my side, it's ready.

(The only still annoying things are the arrow stuff https://github.com/root-project/root/issues/10209 )

jalopezg-r00t commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -90,42 +96,43 @@ namespace textinput {
-    } else {
-      size_t startAt = fCurHistEntry;
-      if (startAt == (size_t) -1) {
+    ssize_t NewHistEntry = -1;
```
See the comment above; this should work, I think.  Apply the same change below for the rest of `ssize_t` occurrences.
```suggestion
    std::ptrdiff_t NewHistEntry = -1;
```

jalopezg-r00t commented on core/textinput/src/textinput/Editor.cpp:
```diff
@@ -90,42 +96,43 @@ namespace textinput {
-      size_t startAt = fCurHistEntry;
-      if (startAt == (size_t) -1) {
+    ssize_t NewHistEntry = -1;
+    if (!fSearch.empty()) {
```
If `fSearch` is empty, the function returns `false` in line 149, which makes `Editor::ProcessCommand()` return `kPRError` (terminal bell ring).

Instead, I would suggest the following - don't forget to remove the closing brace in line 132 and adjust indentation:
```suggestion
    if (fSearch.empty())
      return true;
```

jalopezg-r00t reviewed (changes requested):
Almost there, @ferdymercury!  :slightly_smiling_face:  Many thanks!
Can we squash most of the commits after applying these changes?

jalopezg-r00t commented:
> (The only still annoying things are the arrow stuff #10209 )

I already have a fix for that, but I want to merge this PR first.

ferdymercury commented:
Thanks for the review! I rebased and squashed some of the commits. I guess you can squash the remaining ones while merging ? Or should I do it on my side ?

jalopezg-r00t reviewed (commented):
LGTM!  Well done!  :slightly_smiling_face:

jalopezg-r00t commented:
> > (The only still annoying things are the arrow stuff #10209 )
> 
> I already have a fix for that, but I want to merge this PR first.

See https://github.com/root-project/root/pull/10281.
